### ISSUE TITLE
Incorrect Parsing of YAML Hash with Union Value

### ISSUE THREAD

**Issue by jwaldrip** [crystal-lang/crystal#7936]
When trying to do a YAML.mapping in a class with the unique combination of String, Nil, and another YAML.mapping class `value: null` gets captured as a string "null" and not as nil as it should. null without quotes is a reserved YAML word and should always be treated as such.

Crystal Version 0.29.0

```crystal
require "yaml"

class Foo
  YAML.mapping(
    bar: Hash(String, String | Nil | Raz)
  )
end

class Raz
  YAML.mapping(
    taz: String?
  )
end

yaml = <<-YAML
bar: 
  baz: null
YAML

foo = Foo.from_yaml(yaml)

puts foo
# => {"baz" => "null"}
```

Also, when the value is empty in yaml, it defaults to the first empty class.

```crystal
require "yaml"

class Foo
  YAML.mapping(
    bar: Hash(String, String | Nil | Raz)
  )
end

class Raz
  YAML.mapping(
    taz: String?
  )
end

yaml = <<-YAML
bar: 
  baz:
YAML

foo = Foo.from_yaml(yaml)

puts foo
# => {"baz" => #<Raz:0x103f7e880 @taz=nil>}
```

**Comment by jwaldrip** [comment]
As a note. I would like to point out that there was a fix in this back when I wrote #4838. But a core dev decided to take a different approach. I will dig in and see if I can implement a fix. In the meantime, I would love another set of eyes.

**Comment by asterite** [comment]
@jwaldrip I don't understand the problem in the second example you mention. The compiler has no idea what type you would like to parse first.

My advice is: in case of ambiguity, define a custom converter. I wouldn't suggest using unions in YAML/JSON mapping. Though with #7938 the situation will be a bit better.
